core/src/CMakeLists.txt

```diff
@@ -69,6 +69,9 @@ ELSE()
   LIST(REMOVE_ITEM KOKKOS_CORE_SRCS ${CMAKE_CURRENT_SOURCE_DIR}/impl/Kokkos_Serial_task.cpp)
 ENDIF()
 
+FOREACH (PLUGIN ${Kokkos_PLUGIN_PATH})
+  INCLUDE(${PLUGIN}/core/src/CMakeLists.txt)
```

Comment:
High-level design question:

Do we want the plugins to be built into kokkoscore? Or do we want to make them their own targets outside of core? We can make a plugin a part of `Kokkos::kokkos` without making it a part of `kokkoscore`. That might be a bit of a cleaner separation?

I think the basic motivation here is that you don't want people to have to install and link against external components. You just make a single Kokkos install. At the same time, it's easier from a build-system perspective to build plugins as external things linking against an installed kokkoscore.  I think we can get the best of both worlds with the following workflow:

1. Build kokkoscore same as before
2. Have an option -DKokkos_PLUGINS=A,B,C
3. Loop through each option and include project `add_subdirectory(${Kokkos_PLUGIN_PATH_A})`
4. Each plugin project can just link to kokkoscore. In fact, each project can build not knowing if kokkoscore is in-tree or installed or if it is a subproject or a standalone build. 
5. The plugin can be exported in KokkosConfig.cmake and made a part of `INTERFACE_LINK_LIBRARIES` for Kokkos::kokkos.

Reply:
Does this workflow address the flowing requirements?
1. User is able to include single Kokkos_Core.hpp regardless of what plugins are enabled?
2. User can override default memory space i.e. -DKokkos_DEFAULT_HOST_MEMORY_SPACE=xyz or -DKokkos_DEFAULT_DEVICE_MEMORY_SPACE=xyz (note that this has to work with existing - builtin execution spaces)
3. User can override default Execution Space -DKokkos_DEFAULT_HOST_PARALLEL_EXECUTION_SPACE=xyz ...

Also, i'm pretty sure with the add_subdirectory(...) method, each plugin would have to create a separate set of unit test apps for each backend that is possible/enabled. Note that this is minor compared to the other 3.

before we make and design changes I want to be really confident of the feasibility.  We have a working solution and have to decide if further experimentation is worth the effort.  thanks.

Reply:
Couple questions here:

```cmake
FOREACH (PLUGIN ${Kokkos_PLUGIN_PATH})
  INCLUDE(${PLUGIN}/core/src/CMakeLists.txt)
```

Why not just:

```cmake
INCLUDE(${KOKKOS_PLUGIN})
```

And the user just sets `-DKokkos_PLUGIN=/path/to/some/cmake/script`? If there are multiple plugin folders, the user can just do their own includes in that file. It would help eliminate the user having to use a very specific directory structure.

Also, instead of things like this:

```cmake 
FOREACH (PLUGIN ${Kokkos_PLUGIN_PATH})
  KOKKOS_LIB_INCLUDE_DIRECTORIES(kokkoscore
  ${PLUGIN}/core/src)
ENDFOREACH()
```

I would think we should create an INTERFACE library `kokkosplugin` and user can just add to that target, e.g.

```cmake
ADD_LIBRARY(kokkosplugin INTERFACE)
INCLUDE(${KOKKOS_PLUGIN})
KOKKOS_LINK_TPL(kokkoscore PUBLIC kokkosplugin)
```

and in the `${KOKKOS_PLUGIN}`:

```cmake
target_sources(kokkosplugin INTERFACE ...)
target_include_directories(kokkosplugin INTERFACE ...)
# ... etc.
```

and if there is some weird tribits thing, you can loop through the properties of the target and get the underlying includes, flags, link libs, etc.

Reply:
this makes sense.  I will have to investigate if this solution will accomplish what we need. following the requirements stated above